processing/src/main/java/io/druid/query/lookup/LookupReferencesManager.java

```diff
@@ -52,333 +55,400 @@
-  private final ConcurrentMap<String, LookupExtractorFactory> lookupMap = new ConcurrentHashMap<>();
-  // This is a lock against the state of the REFERENCE MANAGER (aka start/stop state), NOT of the lookup itself.
-  private final ReadWriteLock startStopLock = new ReentrantReadWriteLock(true);
-  private final AtomicBoolean started = new AtomicBoolean(false);
+  private static final EmittingLogger LOG = new EmittingLogger(LookupReferencesManager.class);
+
+  // Lookups state (loaded/to-be-loaded/to-be-dropped etc) is managed by immutable LookupUpdateState instance.
+  // Any update to state is done by creating updated LookupUpdateState instance and atomically setting that
+  // into the ref here.
+  // this allows getAllLookupsState() to provide a consistent view without using locks.
+  @VisibleForTesting
+  final AtomicReference<LookupUpdateState> stateRef = new AtomicReference<>();
+
+  @VisibleForTesting
+  final LookupSnapshotTaker lookupSnapshotTaker;
+
+  @VisibleForTesting
+  final LifecycleLock lifecycleLock = new LifecycleLock();
+
+  @VisibleForTesting
+  Thread mainThread;
 
-  private final LookupSnapshotTaker lookupSnapshotTaker;
+  //for unit testing only
+  private final boolean testMode;
 
   @Inject
-  public LookupReferencesManager(LookupConfig lookupConfig, final @Json ObjectMapper objectMapper)
+  public LookupReferencesManager(LookupConfig lookupConfig, @Json ObjectMapper objectMapper)
+  {
+    this(lookupConfig, objectMapper, false);
+  }
+
+  @VisibleForTesting
+  LookupReferencesManager(LookupConfig lookupConfig, ObjectMapper objectMapper, boolean testMode)
   {
     if (Strings.isNullOrEmpty(lookupConfig.getSnapshotWorkingDir())) {
       this.lookupSnapshotTaker = null;
     } else {
       this.lookupSnapshotTaker = new LookupSnapshotTaker(objectMapper, lookupConfig.getSnapshotWorkingDir());
     }
+
+    this.testMode = testMode;
   }
 
   @LifecycleStart
   public void start()
   {
-    startStopLock.writeLock().lock();
+    if (!lifecycleLock.canStart()) {
+      throw new ISE("can't start.");
+    }
+
     try {
-      if (!started.getAndSet(true)) {
-        if (lookupSnapshotTaker != null) {
-          final List<LookupBean> lookupBeanList = lookupSnapshotTaker.pullExistingSnapshot();
-          for (LookupBean lookupBean : lookupBeanList) {
-            this.put(lookupBean.name, lookupBean.factory);
-          }
-        }
-        LOGGER.info("Started lookup factory references manager");
+      LOG.info("LookupReferencesManager is starting.");
+
+      loadSnapshotAndInitStateRef();
+
+      if (!testMode) {
+        mainThread = Execs.makeThread(
+            "LookupReferencesManager-MainThread",
+            new Runnable()
+            {
+              @Override
+              public void run()
+              {
+                try {
+
+                  if (!lifecycleLock.awaitStarted()) {
+                    LOG.error("WTF! lifecycle not started, lookup update notices will not be handled.");
```

Comment:
not really, that should never happen unless there is a coding error somewhere.
also, in case, it does happen then coordinator would not be able to work with this node and raise alerts.